tests/parametric/test_span_links.py

```diff
@@ -0,0 +1,189 @@
+                    ["x-datadog-parent-id", "9876543210"],
+                    ["x-datadog-sampling-priority", "2"],
+                    ["x-datadog-origin", "synthetics"],
+                    ["x-datadog-tags", "_dd.p.dm=-4,_dd.p.tid=0000000000000010"],
+                ],
+            ):
+                pass
+
+        span = get_span(test_agent)
+        assert span.get("trace_id") != 1234567890
+        assert span_has_no_parent(span)
+        assert span["meta"].get(ORIGIN) is None
+        assert span["meta"].get("_dd.p.dm") != "-4"
+        assert span["metrics"].get(SAMPLING_PRIORITY_KEY) != 2
+
+        assert span.get("span_links") is not None
+
+        span_links = span["span_links"]
+        assert len(span_links) == 1
+        link = span_links[0]
+        assert link.get("span_id") == 9876543210
+        assert link.get("trace_id") == 1234567890
+        assert link.get("trace_id_high") == 16
+
+        assert link.get("tracestate") is not None
+        tracestateArr = link["tracestate"].split(",")
+        assert len(tracestateArr) == 1 and tracestateArr[0].startswith("dd=")
+        tracestateDD = tracestateArr[0][3:].split(";")
+        assert len(tracestateDD) == 3
+        assert "o:synthetics" in tracestateDD
+        assert "s:2" in tracestateDD
+        assert "t.dm:-4" in tracestateDD
+
+        assert (link.get("flags") or 0) == 0
+        assert len(link.get("attributes")) == 2
+        assert link["attributes"].get("foo") == "bar"
+        assert link["attributes"].get("_dd.p.dm") == "-4"
+
+    def test_span_link_from_distributed_w3c_headers(self, test_agent, test_library):
+        """Properly inject w3c distributed tracing information into span links.
+        This mostly tests that the injected tracestate and flags are accurate.
+        """
+        with test_library:
+            with test_library.start_span(
+                "root",
+                links=[Link(parent_id=0)],
+                http_headers=[
+                    ["traceparent", "00-12345678901234567890123456789012-1234567890123456-01"],
+                    ["tracestate", "foo=1,dd=t.dm:-4;s:2"],
+                ],
+            ):
+                pass
+
+        span = get_span(test_agent)
+        assert span.get("trace_id") != 1234567890
+        assert span_has_no_parent(span)
+        assert span["meta"].get("_dd.p.dm") != "-4"
+        assert span["metrics"].get(SAMPLING_PRIORITY_KEY) != 2
+
+        assert span.get("span_links") is not None
+
+        span_links = span["span_links"]
+        assert len(span_links) == 1
+        link = span_links[0]
+        assert link.get("span_id") == 1311768467284833366
+        assert link.get("trace_id") == 8687463697196027922
+        assert link.get("trace_id_high") == 1311768467284833366
+
+        assert link.get("tracestate") is not None
+        tracestateArr = link["tracestate"].split(",")
+        assert len(tracestateArr) == 2 and tracestateArr[0].startswith("dd=")
```

Comment:
This seems incorrect to me. I think it should be the incoming value of `"foo=1,dd=t.dm:-4;s:2"`

Reply:
However, the tracestate you asserted would be fine for the `span` since you are merging the tracestate into the local span

Reply:
Actually I think I see why you've set this as `"foo=1,dd=t.dm:-4;s:2"` I think it is because of this text from the RFC:
> TraceState [optional]: The span’s tracestate field, as defined in the [W3C standard](https://www.w3.org/TR/trace-context/#tracestate-header).
> - If datadog (or b3) headers are given, they must be transformed to an equivalent tracestate representation

I'm a little conflicted on this. It looks like what we are trying to do is to extract the headers and then _inject_ the resulting tracestate into the span link tracestate. I think it would be more appropriate to literally copy over the incoming tracestate header and if we extract context with another propagator then the span link tracestate will be empty. What are your thoughts on this?

Reply:
I think it might be fine to say that the order does not matter.
Ultimately it should not matter much for the purposes of our backend.

And yes, this is the result of extract+inject, because implementation wise, to get the datadog/b3->tracestate transformation, the easiest way is just to extract and inject it, for every language I suppose. Otherwise it will require quite some custom logic, which has no benefit here.

Reply:
Changed :-)

Reply:
Sounds good to me, thanks for the changes!